src/main/java/com/ably/kafka/connect/offset/OffsetRegistryService.java

```diff
@@ -0,0 +1,75 @@
+package com.ably.kafka.connect.offset;
+
+import com.google.common.annotations.VisibleForTesting;
+import org.apache.kafka.clients.consumer.OffsetAndMetadata;
+import org.apache.kafka.common.TopicPartition;
+import org.apache.kafka.connect.sink.SinkRecord;
+
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+
+/**
+ * This class is responsible for keeping track of the latest offset for each topic partition.
+ */
+public class OffsetRegistryService implements OffsetRegistry {
+
+    // Hashmap of TopicPartition to Offsets.
+    private final Map<TopicPartition, Long> topicPartitionToOffsetMap = new ConcurrentHashMap<>();
+
+    public OffsetRegistryService() {
+
+    }
+
+    /**
+     * Updates the topicPartitionToOffsetMap with the latest offset for each topic partition.
+     * @param records
+     */
+    @Override
+    public void updateTopicPartitionToOffsetMap(List<SinkRecord> records) {
+        for (SinkRecord record : records) {
+            TopicPartition topicPartition = new TopicPartition(record.topic(), record.kafkaPartition());
+
+            if (topicPartitionToOffsetMap.containsKey(topicPartition)) {
+                topicPartitionToOffsetMap.compute(topicPartition, (k, v) -> {
+                  if(v < record.kafkaOffset()) {
+                      return record.kafkaOffset();
+                  } else {
+                      return v;
+                  }
+                });
+            } else {
+                topicPartitionToOffsetMap.put(topicPartition, record.kafkaOffset());
```

Comment:
There's still a race here because you mutate `topicPartitionToOffsetMap` differently depending on the return value of `containsKey()`, and those are separate operations that are not guaranteed to be atomic. e.g:

```
Thread 1: topicPartitionToOffsetMap.containsKey("partition_1") // false
Thread 2: topicPartitionToOffsetMap.containsKey("partition_1") // false
Thread 1: topicPartitionToOffsetMap.put("partition_1", 11) 
Thread 2: topicPartitionToOffsetMap.put("partition_1", 10) // wrong -- we overwrite a higher offset 
```

The docs for `compute()` suggest you don't need this `containsKey()` check anyway, as it will just pass `null` as the current value if it's not currently set:

> Attempts to compute a mapping for the specified key and its current mapped value (or null if there is no current mapping).

So, I think you can fix the race by simplifying this code to just:

```java
topicPartitionToOffsetMap.compute(topicPartition, (k, v) -> {
  if(v == null || v < record.kafkaOffset()) {
    return record.kafkaOffset();
  } else {
    return v;
  }
});
```